trie/database_wrap.go

```diff
@@ -0,0 +1,267 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"errors"
+	"runtime"
+	"time"
+
+	"github.com/VictoriaMetrics/fastcache"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/trie/triedb/hashdb"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+)
+
+// Config defines all necessary options for database.
+type Config struct {
+	Cache     int    // Memory allowance (MB) to use for caching trie nodes in memory
+	Journal   string // Journal of clean cache to survive node restarts
+	Preimages bool   // Flag whether the preimage of trie key is recorded
+}
+
+// backend defines the methods needed to access/update trie nodes in different
+// state scheme.
+type backend interface {
+	// Update performs a state transition by committing dirty nodes contained
+	// in the given set in order to update state from the specified parent to
+	// the specified root.
+	Update(root common.Hash, parent common.Hash, nodes *trienode.MergedNodeSet) error
+
+	// Commit writes all relevant trie nodes belonging to the specified state
+	// to disk. Report specifies whether logs will be displayed in info level.
+	Commit(root common.Hash, report bool) error
+
+	// Initialized returns an indicator if the state data is already initialized
+	// according to the state scheme.
+	Initialized(genesisRoot common.Hash) bool
```

Comment:
Nitpicks, could we reorder the methods based on lifecycle?

I.e.

- Scheme
- Initialized
- Size
- Update
- Commit
- Close

Also, for Initialized, could you add a short memo here as to where it's used for? Seems like an extremely specific thing to have a dedicated method checking a "genesis" root which requires the caller to pass the correct hash.

Since there's nothing internally or externally enforcing that the user really passes in a genesisRoot vs something else, could we perhaps have this method called "Exists" or something?

Reply:
Re initialized, it's used to determine if the state in disk is already initialized or not. For hash-based, we only need to check the genesis state root, whereas for path-based, it's different because genesis state may already be pruned.

So we can't simply change it to `Exists` as we don't know which state to check in path-based.

Reply:
Ok, I'll wait for the pbss stuff and will recorsider it when it lands.